### ISSUE TITLE
*Questions* Connectivity to the car

### ISSUE THREAD

**Issue by estingala** [rampage128/niscan#4]
Hi!

First of all, thank you for providing your work to the community!. I own a 370z (GT - 2015 - european model) and plan to replace the very outdated infotainment with a tablet. 

I have chosen a USB-C tablet I have for work (a Galaxy S5e), which seemed like a good choice given its compatibility with usb-c hubs with power delivery, and its very good screen. I plan to add an EasyCap device for the (built-in) rearview camera as well as a CarPlay/Android auto dongle. I'm probably planning to leave the factory AC controls, which I believe will fit below the tablet. I've been investigating around the 370z service manuals and have everything almost figured out, but seeing your set-up (here and in your YouTube video) I have a couple of questions that you have have already stumbled upon:

- You mentioned in another issue that the AC control unit would be incompatible with another redundant control. Is there any particular reason for this?
- The Bose amp takes 4 (balanced) inputs. I assume you have wired the negative part of the signal to ground. Did this produce unwanted noise (alternator whinning or something else)?
- The connectors seem to be proprietary. As I want a non-destructive set-up, did you find these connectors readily available?

I've tried your software in a rig I have for testing and seems to be working great. I also like a lot the way you've organized the code. It has provided me a lot of inspiration!
Thanks again and I hope to share also my progress in this endeavour :)

**Comment by rampage128** [comment]
Hey,

Sounds like a nice plan you have there. Good luck with your build.

Regarding keeping the OEM A/C dials:

Unfortunately for NAV versions of the 370z, the A/C-dials do **not** work without the OEM navigation unit. This is because the dials are not directly communicating with the A/C-amp. They actually communicate with the navigation unit, which then relays the values to the A/C-amp. AFAIK there is no way around this other than using the non-nav dials and connecting those to the arduino to read the dial state.

Regarding your questions:

- The AC-control unit communicates with the navigation system. This communication is tightly coupled over the can bus interface of the car. That means the A/C-Amp sends its status in various packages continuously and the navigation system sends the desired user input continuously. If you keep the navigation system connected and add another device that sends its own packages to the can bus, these will overlap and confuse the AC-Amp. Besides that, even if you would go with plan B from above (using the non nav dial on the arduino), it would logically mean that you have two sources of truth for the dial settings (one physical and one virtual). If you change one of the virtual sliders in the app to a distinct state, this will not match the setting of the physical dial. So the app would have to be in a "read only" mode. This is currently not supported. I hope I explained these two issues in an understandable manner. 
- Yes, I connected all the negative wires to the ground of the headphone jack and the left and right channels to the channels of the headphone jack. At first I indeed had all kinds of annoying noise in certain situations (like android not playing any sound for longer than 5 seconds). Since I switched out the power supply and added some capacitors to the PCB, it runs without any noise except the electric folding of the mirrors. This one produces some barely audible scratchy ticking noise I can not get rid of.
See U2 and C1-C4 in [the current PCB](https://easyeda.com/rampage-128/carduino-v2) for reference.
- I just ordered an ISO radio adapter and cut off the ISO plugs to connect them to my PCB. You will want to get a fully wired adapter. Parrot offers "loop through" adapters for their devices which have all pins connected. If you use a "normal" ISO adapter, it will not have all the pins connected.

**Comment by estingala** [comment]
- Regarding the NAV AC controls, that's a pity. The NAV version of it is stateless and would make for a very nice physical addition _if_ it would be usable. 

- I had past negative experience with unbalanced signals and car amplifiers, where alternator whine and generally every electric actuator would produce unwanted noise through the amplifier. This is in my belief due to ground loops and basically the ground not being stable enough. In the past I had a Nexus 7 in my past car, coupled with a cheap amplifier that made me cry every time I put the volume on :). As we have access to differential signal amplifier, have you tried to to build into your PCB a single ended to differential converter? That may work to completely remove the unwanted noise and only requires a couple of op-amps. In my build, I was planning to use a class compliant USB sound card with built-in balanced outputs (just to avoid the noise-related headaches) such as the Behringer UMC22, as my tablet doesn't include an analog output and would need an external one anyway.

- Thanks for the tip! I'm actually in the very early (theoretical stage) of the project, and I haven't actually opened the guts of the car yet. I'm glad to hear that there's an ISO connector in there :)

I'm also on EcuTek and would like to build at least some more extra functionality by interfacing ECM signals (speed, load, ignition timing, current map setting). I saw that your current functionality is consuming around half of the serial rate speed, but see that you still have plenty of room to increase baud rate of the serial port. Have you seen the Arduino as a bottleneck in terms of processing power to process your current set-up? Do you think it's a viable platform to build more features on top or should I consider alternatives like the ESP32?

Another thing is the mechanical aspect of the build. I plan to build my own using my 3d printer. Did you do the same? I'm suspecting it will be good for prototyping, but long term usage will require of a more heat-resistant material.

I'm excited about this build and hope to share as much as possible!

**Comment by rampage128** [comment]
- Actually I never checked how the OEM NAV dials work. I know they are stateless (except for the LEDs). So maybe you can actually try to read those using the arduino and also control the LEDs using the arduino. The carduino project PCB has expansion slots on the mainboard. Maybe we could design an expansion board for that purpose.
- Actually it might also depend on the tablet quality. With the first tablet I used, you could actually hear high pitch wining and faint "modem-like" sounds when the CPU of the tablet was working. I was planning to design an adjustable pre-amp that would also allow to balance the left/right, front/back channels. It could be plugged into the main-board as an extension module. The goal was to have a line in for the tablet and a connector for the car speakers. And being able to control the balancing from the tablet. But I have no clue about audio amplifiers or how to build one.
- I think I didn't express that clearly. The car has proprietary connectors. But you can get adapters to ISO online. Sadly the normal ISO radio adapters do not use all the pins on the plug, so you will want to buy one from parrot ([See this one for reference](https://www.amazon.de/gp/product/B01JZ0XZAW/)). For my build I cut off the ISO-End and crimped on molex connectors and the audio cable to effectively create my own adapter.
- The arduino can be a bottleneck. Especially the nano knockoffs. But CPU power was not an issue yet. The serial connection is the real bummer. The knockoffs use CH340 USB serial chips. These have a limited baud rate. I think 115200 is the limit. According to the datasheet more should be possible, but I could not get it to work reliably. I am currently working on custom layouts with meters (like torque allows). Then the user can create screens and put them into the home screen or start them from custom icons. This would then also allow the user to customize the UI of the A/C controls. But I have very limited time and that project is quite complex. My guess is that after adding RPM/Speed and other realtime data, either serial bandwidth or memory of the arduino will be a bottleneck. If you want to be on the safe side, you can switch to teensy. They use full USB-speed for the baud rate, have much more power and memory and also allow you to connect them as serial/keyboard/mouse at the same time. Especially the keyboard feature is nice, because it allows you to access some android functions which can not be done from an app (like switching tasks, emulating the back button and so on).
- For the case of the PCB I will use an ABS 3D print. That should easily resist the temperatures inside the dashboard. For the frame of the tablet I am actually struggling. I started 3 times already to build something out of GRP. The issue is the ridiculous shape of the opening in the dashboard. I just can not get it right/find the right method to create a frame. And for 3D printing I would need a 3D scan of the outer frame. There is none available so printing is out of the question.

**Comment by xadamadax** [comment]
Hey guys, I'm actually trying to accomplish this in a GTR.  the only spot for an aftermarket stereo is where the AC controls are. 

Am I understanding correctly that this "will not" work in the Z that has the nav unit?  The GTR "only" comes with that unit, so I'm trying too gauge my likelihood of success before jumping in.

This would be game changing for the GTR community as  many are stuck with the 10 year old head unit.

**Comment by rampage128** [comment]
Hey!

I have a 370z with nav. This project only works *with* the nav unit. Since only with the nav unit the AC can be controlled via can-bus.

I don't know the GTR specifics tho. It could be that the AC works differently in the GTR. But with hands on a GTR that could be figured out easily.

**Comment by xadamadax** [comment]
Thanks so much for the response!  I have extensive background in AV and video game modding, I run cyber security practice for a living.  Arduino however is brand new to me.  I have read a bit already but there is so much i am looking to target what I need.

If you could help point me at the right resources - or better still if you'd be willing to collaborate.  There is a potential market for the GTR to have a stereo solution that doesn't require hacking up the AC.  Particularly as all vehicles from 2009 to 2016 have a garbage infotainment - which i believe is identical to the unit in your z.

I picture an android headunit with this integrated via usb

**Comment by rampage128** [comment]
Sure, I would love to collaborate.

Currently the requirements to use this project are:

- An android tablet that supports USB-OTG and simultaneous charging or any android car stereo with USB ports.
- The hardware found on https://easyeda.com/rampage-128/carduino-v2
- The firmware found on https://github.com/rampage128/carduino/tree/library-preparation
(Note that this does the same as the niscan project we are commenting on but has a newer version. carduino will later turn into a library and this project (niscan) will be the nissan implementation. So currently it is better to use carduino)
- The Android app found on https://github.com/rampage128/cardroid/tree/architecture/maintenance-update
(You will have to build it yourself using android studio)
- A radio adapter for the GTR (with all pins, some adapters are missing some pins). The ISO-End of the adapter will be replaced with ATX plugs that match the carduino hardware.
- A micro usb extension cord (with all 5 pins). One end will be cut off and connected to the carduino ATX plug, the other end goes into the tablet.
- A headphone extension cord or Y-Adapter to connect to the speaker wires of the radio adapter.
- Depending on your audio system (and if you use a tablet) you might need a preamp to amplify the audio signal from the android device. In the 370z with the bose system no preamp is needed. I don't know about the GTR.

I would love to contribute more, since I hope that adding the GTR might involve more people into development of this system. Unfortunately I have no GTR and I do not have access to one. I don't even know anyone who owns a GTR. This means you would have to build and test the stuff on your own to find out if it works. If it doesn't I will be happy to assist in fixing any potential issues.

An alternative would be if you have access to a can-sniffer. Then you could check the can bus and we could do a preliminary check for compatibility. Mainly to see if the packet-ids in the GTR match the 370z.

**Comment by xadamadax** [comment]
Awesome!!  I think can sniffing will be beneficial regardless to see what additional functionality we want to add.  We have access to more than few GTR and several tablets and android double dins 

Wasn't sure the best way to exchange contact info, so i'll just put my email here so we can move communication outside of your bugs section :)

xadamadax@gmail.com

Shoot me a line and we can do a webex and do some brain storming/create like a project plan.  I'll start looking through the links you sent over- thanks again for reaching back out, looking forward to it

Adam

**Comment by estingala** [comment]
Hi! 

I now have a working design for a slide-in solution for a Galaxy Tab s5e (https://imgur.com/a/IqEHV05). I've printed this with Petg and it has (so far) withstood the summer in the Canary Islands :). As we commented before, my set-up differs a bit of what the Carduino hardware offers, and I had to build my own solution for this.

I would love to collaborate in the project. I'm highly skilled in embedded systems, C, C++ and currently learning Android (I do audio development for iOS, macOS and embedded for a living). Perhaps we can set-up a Discord or Slack group where we can sync on our different goals and ambitions and align them together?

**Comment by rampage128** [comment]
Great frame! I have very good news for you regarding reading ECM data. I am currently working on an update of firmware and app. All the data you want to retrieve is now directly configurable in the app. No need to update the firmware on the arduino for new values anymore.

It already works, I just have to finish the user interface and iron out some bugs to make it ready to be pushed and merged.

Would you be willing to share the 3D file for the frame? Then I could make a version for the trekstor tablet. 

The slack group is a great idea! I will create one. Feel free to drop me an email (frederik.wolter@gmail.com)!

**Comment by h17nan** [comment]
hi i have a 370z nav and want to make a double android screen setup. i have little understand on how to get your setup to work. 
what i have done so far 
purchased mcp2515 and arduino nano
cloned and uploaded on to the arduin. niscan(i just couldnt figure out how to get carduino to work), AnalogMultiButton, everytime, MCP_CAN_lib, arduino - mcp2515- master. 
created the app using android studios. and have installed it on my phone(s10)
i have done all the pins to connect nano to mcp and connected to h line and L line of car. 

the issue is when i connect arduino to my phone it seems to be connected however nothing shows up in the app.
i know its hard to see where im going wrong but is there anyway u could dumb it down for me or tell me what im doing wrong
